core/src/main/scala/chisel3/internal/SourceInfo.scala

```diff
@@ -43,17 +44,38 @@ case object DeprecatedSourceInfo extends NoSourceInfo
 
 /** For FIRRTL lines from a Scala source line.
   */
-case class SourceLine(filename: String, line: Int, col: Int) extends SourceInfo {
-  def makeMessage(f: String => String): String = f(s"@[$filename $line:$col]")
+case class SourceLine(fullPath: String, line: Int, col: Int) extends SourceInfo {
+  val tokens = fullPath.split('/').reverse
+  val computor = chisel3.internal.Builder.sourcePathComputor
+  computor.map(_.addPath(tokens))
+  lazy val relativeFileName = computor.map(_.getFileName(tokens)).getOrElse(tokens.head)
+  def makeMessage(f: String => String): String = f(s"@[$relativeFileName $line:$col]")
 }
 
+object SourcePathComputor {
+  def empty() = SourcePathComputor(Trie.empty[String, Unit])
+}
+case class SourcePathComputor(trie: Trie[String, Unit]) {
+  def getFileName(path: Seq[String]): String = {
+    def getMinimalPath(t: Trie[String, Unit], p: Seq[String]): Seq[String] = {
+      if(t.getChildren.size == 1) Nil else {
+        val childTrie = t.getChildren.collectFirst{ case (token, childTrie) if token == p.head => childTrie}.get
+        p.head +: getMinimalPath(childTrie, p.tail)
+      }
+    }
+    getMinimalPath(trie, path).reverse.mkString("/")
+  }
+  def addPath(path: Seq[String]): Unit = trie.insert(path, ())
+}
+
+
 /** Provides a macro that returns the source information at the invocation point.
   */
 object SourceInfoMacro {
   def generate_source_info(c: Context): c.Tree = {
     import c.universe._
     val p = c.enclosingPosition
-    q"_root_.chisel3.internal.sourceinfo.SourceLine(${p.source.file.name}, ${p.line}, ${p.column})"
+    q"_root_.chisel3.internal.sourceinfo.SourceLine(${p.source.file.canonicalPath}, ${p.line}, ${p.column})"
```

Comment:
A sketch of the solution I'm proposing:

```suggestion
    val root = sys.env.get("chisel_project_root").getOrElse(/* Figure out what to do if not provided*/
    val path = p.source.file.canonicalPath.stripPrefix(root) // TODO warn/error if nothing is stripped
    q"_root_.chisel3.internal.sourceinfo.SourceLine($path, ${p.line}, ${p.column})"
```

Then the user in their `build.sbt` just needs something like:
```scala
Compile / javaOptions += "-Dchisel_project_root=" + baseDirectory.value
```
And I think that's basically all we need.